<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.8.26">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>10&nbsp; Simulation &amp; Parentage â€“ gstudio: Spatial Analysis of Genetic Marker Data</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./visualization.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js" type="module"></script>
<script src="site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="site_libs/quarto-html/axe/axe-check.js" type="module"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-597958c53c93a607afca12fd375c57ed.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-25a4e8620ae17d49cd0c3861fdf92567.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-sidebar floating quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./simulation-parentage.html"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Simulation &amp; Parentage</span></a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">gstudio: Spatial Analysis of Genetic Marker Data</a> 
        <div class="sidebar-tools-main">
    <a href="https://github.com/dyerlab/gstudio" title="Source Code" class="quarto-navigation-tool px-1" aria-label="Source Code"><i class="bi bi-github"></i></a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preface</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./getting-started.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Getting Started</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./locus-class.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">The Locus Class</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./data-io.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Importing &amp; Exporting Data</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./allele-frequencies.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Allele Frequencies &amp; HWE</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./genetic-diversity.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Genetic Diversity</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./genetic-structure.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Population Structure</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./genetic-distance.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Genetic Distance &amp; Relatedness</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./population-graphs.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Population Graphs</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./visualization.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Visualization</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./simulation-parentage.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Simulation &amp; Parentage</span></span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#building-populations-from-frequencies" id="toc-building-populations-from-frequencies" class="nav-link active" data-scroll-target="#building-populations-from-frequencies"><span class="header-section-number">10.1</span> Building Populations from Frequencies</a>
  <ul class="collapse">
  <li><a href="#allele-frequency-specification" id="toc-allele-frequency-specification" class="nav-link" data-scroll-target="#allele-frequency-specification"><span class="header-section-number">10.1.1</span> Allele Frequency Specification</a></li>
  <li><a href="#creating-populations" id="toc-creating-populations" class="nav-link" data-scroll-target="#creating-populations"><span class="header-section-number">10.1.2</span> Creating Populations</a></li>
  </ul></li>
  <li><a href="#migration-models" id="toc-migration-models" class="nav-link" data-scroll-target="#migration-models"><span class="header-section-number">10.2</span> Migration Models</a>
  <ul class="collapse">
  <li><a href="#island-model" id="toc-island-model" class="nav-link" data-scroll-target="#island-model"><span class="header-section-number">10.2.1</span> Island Model</a></li>
  <li><a href="#d-stepping-stone" id="toc-d-stepping-stone" class="nav-link" data-scroll-target="#d-stepping-stone"><span class="header-section-number">10.2.2</span> 1D Stepping Stone</a></li>
  <li><a href="#d-stepping-stone-1" id="toc-d-stepping-stone-1" class="nav-link" data-scroll-target="#d-stepping-stone-1"><span class="header-section-number">10.2.3</span> 2D Stepping Stone</a></li>
  <li><a href="#distance-based" id="toc-distance-based" class="nav-link" data-scroll-target="#distance-based"><span class="header-section-number">10.2.4</span> Distance-Based</a></li>
  </ul></li>
  <li><a href="#migration-events" id="toc-migration-events" class="nav-link" data-scroll-target="#migration-events"><span class="header-section-number">10.3</span> Migration Events</a></li>
  <li><a href="#mutation-models" id="toc-mutation-models" class="nav-link" data-scroll-target="#mutation-models"><span class="header-section-number">10.4</span> Mutation Models</a></li>
  <li><a href="#forward-time-simulation" id="toc-forward-time-simulation" class="nav-link" data-scroll-target="#forward-time-simulation"><span class="header-section-number">10.5</span> Forward-Time Simulation</a>
  <ul class="collapse">
  <li><a href="#basic-simulation" id="toc-basic-simulation" class="nav-link" data-scroll-target="#basic-simulation"><span class="header-section-number">10.5.1</span> Basic Simulation</a></li>
  <li><a href="#with-migration" id="toc-with-migration" class="nav-link" data-scroll-target="#with-migration"><span class="header-section-number">10.5.2</span> With Migration</a></li>
  <li><a href="#with-mutation" id="toc-with-mutation" class="nav-link" data-scroll-target="#with-mutation"><span class="header-section-number">10.5.3</span> With Mutation</a></li>
  <li><a href="#with-mixed-mating-selfing" id="toc-with-mixed-mating-selfing" class="nav-link" data-scroll-target="#with-mixed-mating-selfing"><span class="header-section-number">10.5.4</span> With Mixed Mating (Selfing)</a></li>
  <li><a href="#census-snapshots" id="toc-census-snapshots" class="nav-link" data-scroll-target="#census-snapshots"><span class="header-section-number">10.5.5</span> Census Snapshots</a></li>
  <li><a href="#complete-worked-example" id="toc-complete-worked-example" class="nav-link" data-scroll-target="#complete-worked-example"><span class="header-section-number">10.5.6</span> Complete Worked Example</a></li>
  </ul></li>
  <li><a href="#parentage-analysis" id="toc-parentage-analysis" class="nav-link" data-scroll-target="#parentage-analysis"><span class="header-section-number">10.6</span> Parentage Analysis</a>
  <ul class="collapse">
  <li><a href="#mating-and-offspring" id="toc-mating-and-offspring" class="nav-link" data-scroll-target="#mating-and-offspring"><span class="header-section-number">10.6.1</span> Mating and Offspring</a></li>
  <li><a href="#maternal-subtraction" id="toc-maternal-subtraction" class="nav-link" data-scroll-target="#maternal-subtraction"><span class="header-section-number">10.6.2</span> Maternal Subtraction</a></li>
  <li><a href="#fractional-paternity" id="toc-fractional-paternity" class="nav-link" data-scroll-target="#fractional-paternity"><span class="header-section-number">10.6.3</span> Fractional Paternity</a></li>
  <li><a href="#single-parent-exclusion" id="toc-single-parent-exclusion" class="nav-link" data-scroll-target="#single-parent-exclusion"><span class="header-section-number">10.6.4</span> Single-Parent Exclusion</a></li>
  <li><a href="#identifying-bad-parents" id="toc-identifying-bad-parents" class="nav-link" data-scroll-target="#identifying-bad-parents"><span class="header-section-number">10.6.5</span> Identifying Bad Parents</a></li>
  </ul></li>
  <li><a href="#bundled-parentage-datasets" id="toc-bundled-parentage-datasets" class="nav-link" data-scroll-target="#bundled-parentage-datasets"><span class="header-section-number">10.7</span> Bundled Parentage Datasets</a></li>
  </ul>
<div class="toc-actions"><ul><li><a href="https://github.com/dyerlab/gstudio/issues/new" class="toc-action"><i class="bi bi-github"></i>Report an issue</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Simulation &amp; Parentage</span></h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source"><i class="bi"></i> Code</button></div></div>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<p>This chapter covers the forward-time population simulation framework and parentage analysis tools in <code>gstudio</code>.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(gstudio)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr, <span class="at">warn.conflicts =</span> <span class="cn">FALSE</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<section id="building-populations-from-frequencies" class="level2" data-number="10.1">
<h2 data-number="10.1" class="anchored" data-anchor-id="building-populations-from-frequencies"><span class="header-section-number">10.1</span> Building Populations from Frequencies</h2>
<section id="allele-frequency-specification" class="level3" data-number="10.1.1">
<h3 data-number="10.1.1" class="anchored" data-anchor-id="allele-frequency-specification"><span class="header-section-number">10.1.1</span> Allele Frequency Specification</h3>
<p>The simulation tools start with specifying allele frequencies per population and locus. The <code>make_populations()</code> function requires a data.frame with <code>Population</code>, <code>Locus</code>, <code>Allele</code>, and <code>Frequency</code> columns:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>f1 <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">Population =</span> <span class="st">"A"</span>,</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">Locus =</span> <span class="fu">rep</span>(<span class="fu">c</span>(<span class="st">"Loc1"</span>, <span class="st">"Loc2"</span>, <span class="st">"Loc3"</span>), <span class="at">each =</span> <span class="dv">2</span>),</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">Allele =</span> <span class="fu">rep</span>(<span class="fu">c</span>(<span class="st">"01"</span>, <span class="st">"02"</span>), <span class="at">times =</span> <span class="dv">3</span>),</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">Frequency =</span> <span class="fu">c</span>(<span class="fl">0.2</span>, <span class="fl">0.8</span>, <span class="fl">0.3</span>, <span class="fl">0.7</span>, <span class="fl">0.4</span>, <span class="fl">0.6</span>)</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>f2 <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">Population =</span> <span class="st">"B"</span>,</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">Locus =</span> <span class="fu">rep</span>(<span class="fu">c</span>(<span class="st">"Loc1"</span>, <span class="st">"Loc2"</span>, <span class="st">"Loc3"</span>), <span class="at">each =</span> <span class="dv">2</span>),</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">Allele =</span> <span class="fu">rep</span>(<span class="fu">c</span>(<span class="st">"01"</span>, <span class="st">"02"</span>), <span class="at">times =</span> <span class="dv">3</span>),</span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">Frequency =</span> <span class="fu">c</span>(<span class="fl">0.7</span>, <span class="fl">0.3</span>, <span class="fl">0.5</span>, <span class="fl">0.5</span>, <span class="fl">0.6</span>, <span class="fl">0.4</span>)</span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a>f3 <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a>  <span class="at">Population =</span> <span class="st">"C"</span>,</span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a>  <span class="at">Locus =</span> <span class="fu">rep</span>(<span class="fu">c</span>(<span class="st">"Loc1"</span>, <span class="st">"Loc2"</span>, <span class="st">"Loc3"</span>), <span class="at">each =</span> <span class="dv">2</span>),</span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a>  <span class="at">Allele =</span> <span class="fu">rep</span>(<span class="fu">c</span>(<span class="st">"01"</span>, <span class="st">"02"</span>), <span class="at">times =</span> <span class="dv">3</span>),</span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a>  <span class="at">Frequency =</span> <span class="fu">c</span>(<span class="fl">0.4</span>, <span class="fl">0.6</span>, <span class="fl">0.3</span>, <span class="fl">0.7</span>, <span class="fl">0.2</span>, <span class="fl">0.8</span>)</span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true" tabindex="-1"></a>freqs <span class="ot">&lt;-</span> <span class="fu">rbind</span>(f1, f2, f3)</span>
<span id="cb2-20"><a href="#cb2-20" aria-hidden="true" tabindex="-1"></a>freqs<span class="sc">$</span>Population <span class="ot">&lt;-</span> <span class="fu">factor</span>(freqs<span class="sc">$</span>Population, <span class="at">ordered =</span> <span class="cn">TRUE</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
<section id="creating-populations" class="level3" data-number="10.1.2">
<h3 data-number="10.1.2" class="anchored" data-anchor-id="creating-populations"><span class="header-section-number">10.1.2</span> Creating Populations</h3>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>pop <span class="ot">&lt;-</span> <span class="fu">make_populations</span>(freqs, <span class="at">N =</span> <span class="dv">30</span>)</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(pop)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Population ID  Loc1  Loc2  Loc3
1          A  1 02:02 01:02 01:02
2          A  2 01:02 01:02 02:02
3          A  3 02:02 02:02 02:02
4          A  4 02:02 02:02 01:02
5          A  5 02:02 01:02 02:02
6          A  6 02:02 02:02 01:02</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(pop<span class="sc">$</span>Population)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>
 A  B  C 
30 30 30 </code></pre>
</div>
</div>
<p>You can specify an inbreeding level with <code>F</code>:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>pop_inbred <span class="ot">&lt;-</span> <span class="fu">make_populations</span>(freqs, <span class="at">N =</span> <span class="dv">30</span>, <span class="at">F =</span> <span class="fl">0.3</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
</section>
<section id="migration-models" class="level2" data-number="10.2">
<h2 data-number="10.2" class="anchored" data-anchor-id="migration-models"><span class="header-section-number">10.2</span> Migration Models</h2>
<p>The <code>migration_matrix()</code> function creates a K x K migration matrix where rows sum to 1. Element <code>[i,j]</code> is the proportion of population <em>i</em> that originates from population <em>j</em>.</p>
<section id="island-model" class="level3" data-number="10.2.1">
<h3 data-number="10.2.1" class="anchored" data-anchor-id="island-model"><span class="header-section-number">10.2.1</span> Island Model</h3>
<p>Equal migration among all populations:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">migration_matrix</span>(<span class="dv">3</span>, <span class="at">model =</span> <span class="st">"island"</span>, <span class="at">m =</span> <span class="fl">0.05</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>      Pop1  Pop2  Pop3
Pop1 0.950 0.025 0.025
Pop2 0.025 0.950 0.025
Pop3 0.025 0.025 0.950</code></pre>
</div>
</div>
</section>
<section id="d-stepping-stone" class="level3" data-number="10.2.2">
<h3 data-number="10.2.2" class="anchored" data-anchor-id="d-stepping-stone"><span class="header-section-number">10.2.2</span> 1D Stepping Stone</h3>
<p>Migration only between adjacent populations:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">migration_matrix</span>(<span class="fu">c</span>(<span class="st">"A"</span>, <span class="st">"B"</span>, <span class="st">"C"</span>, <span class="st">"D"</span>),</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>                 <span class="at">model =</span> <span class="st">"stepping_stone_1d"</span>, <span class="at">m =</span> <span class="fl">0.1</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>     A    B    C    D
A 0.90 0.10 0.00 0.00
B 0.05 0.90 0.05 0.00
C 0.00 0.05 0.90 0.05
D 0.00 0.00 0.10 0.90</code></pre>
</div>
</div>
</section>
<section id="d-stepping-stone-1" class="level3" data-number="10.2.3">
<h3 data-number="10.2.3" class="anchored" data-anchor-id="d-stepping-stone-1"><span class="header-section-number">10.2.3</span> 2D Stepping Stone</h3>
<p>Grid-based adjacency:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">migration_matrix</span>(<span class="dv">4</span>, <span class="at">model =</span> <span class="st">"stepping_stone_2d"</span>,</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>                 <span class="at">m =</span> <span class="fl">0.1</span>, <span class="at">nr =</span> <span class="dv">2</span>, <span class="at">nc =</span> <span class="dv">2</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>     Pop1 Pop2 Pop3 Pop4
Pop1 0.90 0.05 0.05 0.00
Pop2 0.05 0.90 0.00 0.05
Pop3 0.05 0.00 0.90 0.05
Pop4 0.00 0.05 0.05 0.90</code></pre>
</div>
</div>
</section>
<section id="distance-based" class="level3" data-number="10.2.4">
<h3 data-number="10.2.4" class="anchored" data-anchor-id="distance-based"><span class="header-section-number">10.2.4</span> Distance-Based</h3>
<p>Migration rate decays with distance:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>coords <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">1</span>, <span class="dv">0</span>, <span class="fl">0.5</span>, <span class="dv">1</span>), <span class="at">ncol =</span> <span class="dv">2</span>, <span class="at">byrow =</span> <span class="cn">TRUE</span>)</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="fu">migration_matrix</span>(<span class="dv">3</span>, <span class="at">model =</span> <span class="st">"distance"</span>, <span class="at">m =</span> <span class="fl">0.1</span>, <span class="at">coords =</span> coords)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>          Pop1      Pop2      Pop3
Pop1 0.9000000 0.0527864 0.0472136
Pop2 0.0527864 0.9000000 0.0472136
Pop3 0.0500000 0.0500000 0.9000000</code></pre>
</div>
</div>
</section>
</section>
<section id="migration-events" class="level2" data-number="10.3">
<h2 data-number="10.3" class="anchored" data-anchor-id="migration-events"><span class="header-section-number">10.3</span> Migration Events</h2>
<p>A <code>migration_event</code> pairs a migration matrix with a generation interval, allowing temporal regime changes:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>m1 <span class="ot">&lt;-</span> <span class="fu">migration_matrix</span>(<span class="dv">3</span>, <span class="at">model =</span> <span class="st">"island"</span>, <span class="at">m =</span> <span class="fl">0.05</span>)</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>m2 <span class="ot">&lt;-</span> <span class="fu">migration_matrix</span>(<span class="dv">3</span>, <span class="at">model =</span> <span class="st">"island"</span>, <span class="at">m =</span> <span class="fl">0.20</span>)</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Low migration for generations 1-50</span></span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>evt1 <span class="ot">&lt;-</span> <span class="fu">migration_event</span>(m1, <span class="at">start =</span> <span class="dv">1</span>, <span class="at">end =</span> <span class="dv">50</span>)</span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a><span class="co"># High migration from generation 51 onward</span></span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a>evt2 <span class="ot">&lt;-</span> <span class="fu">migration_event</span>(m2, <span class="at">start =</span> <span class="dv">51</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
<section id="mutation-models" class="level2" data-number="10.4">
<h2 data-number="10.4" class="anchored" data-anchor-id="mutation-models"><span class="header-section-number">10.4</span> Mutation Models</h2>
<p>The <code>mutation_model()</code> function defines how alleles mutate during simulation:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Infinite Alleles Model</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>mm_iam <span class="ot">&lt;-</span> <span class="fu">mutation_model</span>(<span class="at">rate =</span> <span class="fl">0.001</span>, <span class="at">model =</span> <span class="st">"iam"</span>)</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a><span class="co"># K-Allele Model (10 possible states)</span></span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>mm_kam <span class="ot">&lt;-</span> <span class="fu">mutation_model</span>(<span class="at">rate =</span> <span class="fl">0.001</span>, <span class="at">model =</span> <span class="st">"kam"</span>, <span class="at">k =</span> <span class="dv">10</span>)</span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Stepwise Mutation Model</span></span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a>mm_smm <span class="ot">&lt;-</span> <span class="fu">mutation_model</span>(<span class="at">rate =</span> <span class="fl">0.001</span>, <span class="at">model =</span> <span class="st">"smm"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
<section id="forward-time-simulation" class="level2" data-number="10.5">
<h2 data-number="10.5" class="anchored" data-anchor-id="forward-time-simulation"><span class="header-section-number">10.5</span> Forward-Time Simulation</h2>
<p>The <code>simulate_pop()</code> function runs a multi-generation forward-time simulation. Each generation cycle: census, migrate, mate, mutate.</p>
<section id="basic-simulation" class="level3" data-number="10.5.1">
<h3 data-number="10.5.1" class="anchored" data-anchor-id="basic-simulation"><span class="header-section-number">10.5.1</span> Basic Simulation</h3>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>pop <span class="ot">&lt;-</span> <span class="fu">make_populations</span>(freqs, <span class="at">N =</span> <span class="dv">30</span>)</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>result <span class="ot">&lt;-</span> <span class="fu">simulate_pop</span>(pop, <span class="at">ngen =</span> <span class="dv">20</span>, <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(result)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>  ID Population  Loc1  Loc2  Loc3
1  1          A 01:02 01:01 01:02
2  2          A 02:02 01:01 01:01
3  3          A 02:02 01:02 02:02
4  4          A 02:02 01:02 01:02
5  5          A 02:02 01:02 02:02
6  6          A 01:02 01:01 01:01</code></pre>
</div>
</div>
</section>
<section id="with-migration" class="level3" data-number="10.5.2">
<h3 data-number="10.5.2" class="anchored" data-anchor-id="with-migration"><span class="header-section-number">10.5.2</span> With Migration</h3>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>m <span class="ot">&lt;-</span> <span class="fu">migration_matrix</span>(<span class="dv">3</span>, <span class="at">model =</span> <span class="st">"island"</span>, <span class="at">m =</span> <span class="fl">0.05</span>)</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>evt <span class="ot">&lt;-</span> <span class="fu">migration_event</span>(m, <span class="at">start =</span> <span class="dv">1</span>)</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>result <span class="ot">&lt;-</span> <span class="fu">simulate_pop</span>(pop, <span class="at">ngen =</span> <span class="dv">20</span>,</span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>                       <span class="at">migration =</span> evt,</span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a>                       <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
<section id="with-mutation" class="level3" data-number="10.5.3">
<h3 data-number="10.5.3" class="anchored" data-anchor-id="with-mutation"><span class="header-section-number">10.5.3</span> With Mutation</h3>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>mm <span class="ot">&lt;-</span> <span class="fu">mutation_model</span>(<span class="at">rate =</span> <span class="fl">0.01</span>, <span class="at">model =</span> <span class="st">"iam"</span>)</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>result <span class="ot">&lt;-</span> <span class="fu">simulate_pop</span>(pop, <span class="at">ngen =</span> <span class="dv">20</span>,</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>                       <span class="at">mutation =</span> mm,</span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a>                       <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
<section id="with-mixed-mating-selfing" class="level3" data-number="10.5.4">
<h3 data-number="10.5.4" class="anchored" data-anchor-id="with-mixed-mating-selfing"><span class="header-section-number">10.5.4</span> With Mixed Mating (Selfing)</h3>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>result <span class="ot">&lt;-</span> <span class="fu">simulate_pop</span>(pop, <span class="at">ngen =</span> <span class="dv">20</span>,</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>                       <span class="at">selfing_rate =</span> <span class="fl">0.3</span>,</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>                       <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
<section id="census-snapshots" class="level3" data-number="10.5.5">
<h3 data-number="10.5.5" class="anchored" data-anchor-id="census-snapshots"><span class="header-section-number">10.5.5</span> Census Snapshots</h3>
<p>Save population snapshots at regular intervals:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>result <span class="ot">&lt;-</span> <span class="fu">simulate_pop</span>(pop, <span class="at">ngen =</span> <span class="dv">100</span>,</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>                       <span class="at">census_interval =</span> <span class="dv">25</span>,</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>                       <span class="at">census_dir =</span> <span class="fu">tempdir</span>(),</span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>                       <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Snapshots saved as gen_0025.rds, gen_0050.rds, etc.</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
<section id="complete-worked-example" class="level3" data-number="10.5.6">
<h3 data-number="10.5.6" class="anchored" data-anchor-id="complete-worked-example"><span class="header-section-number">10.5.6</span> Complete Worked Example</h3>
<p>Simulate drift and migration in a 3-population system, then analyze the outcome:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create starting populations</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>pop <span class="ot">&lt;-</span> <span class="fu">make_populations</span>(freqs, <span class="at">N =</span> <span class="dv">50</span>)</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Set up island model migration</span></span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>m <span class="ot">&lt;-</span> <span class="fu">migration_matrix</span>(<span class="dv">3</span>, <span class="at">model =</span> <span class="st">"island"</span>, <span class="at">m =</span> <span class="fl">0.02</span>)</span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a>evt <span class="ot">&lt;-</span> <span class="fu">migration_event</span>(m, <span class="at">start =</span> <span class="dv">1</span>)</span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Run simulation</span></span>
<span id="cb24-9"><a href="#cb24-9" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">42</span>)</span>
<span id="cb24-10"><a href="#cb24-10" aria-hidden="true" tabindex="-1"></a>final <span class="ot">&lt;-</span> <span class="fu">simulate_pop</span>(pop, <span class="at">ngen =</span> <span class="dv">50</span>,</span>
<span id="cb24-11"><a href="#cb24-11" aria-hidden="true" tabindex="-1"></a>                      <span class="at">migration =</span> evt,</span>
<span id="cb24-12"><a href="#cb24-12" aria-hidden="true" tabindex="-1"></a>                      <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb24-13"><a href="#cb24-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-14"><a href="#cb24-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Analyze structure in the final generation</span></span>
<span id="cb24-15"><a href="#cb24-15" aria-hidden="true" tabindex="-1"></a><span class="fu">genetic_structure</span>(final, <span class="at">stratum =</span> <span class="st">"Population"</span>, <span class="at">mode =</span> <span class="st">"Gst"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>       Locus       Gst        Hs        Ht        P
1       Loc1 0.2240740 0.3267340 0.4210891 0.224074
2       Loc2 0.2950070 0.3156229 0.4476965 0.295007
3       Loc3 0.1449410 0.4195286 0.4906429 0.144941
4 Multilocus 0.2188736 1.0618855 1.3594285       NA</code></pre>
</div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Compare diversity before and after</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a><span class="fu">genetic_diversity</span>(pop, <span class="at">stratum =</span> <span class="st">"Population"</span>, <span class="at">mode =</span> <span class="st">"He"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Stratum Locus     He
2        A  Loc1 0.3200
3        A  Loc2 0.4118
4        A  Loc3 0.4800
5        B  Loc1 0.4118
6        B  Loc2 0.5000
7        B  Loc3 0.4800
8        C  Loc1 0.4800
9        C  Loc2 0.4118
10       C  Loc3 0.3200</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="fu">genetic_diversity</span>(final, <span class="at">stratum =</span> <span class="st">"Population"</span>, <span class="at">mode =</span> <span class="st">"He"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Stratum Locus     He
2        A  Loc1 0.0000
3        A  Loc2 0.4712
4        A  Loc3 0.4422
5        B  Loc1 0.4712
6        B  Loc2 0.0000
7        B  Loc3 0.4838
8        C  Loc1 0.4992
9        C  Loc2 0.4662
10       C  Loc3 0.3200</code></pre>
</div>
</div>
</section>
</section>
<section id="parentage-analysis" class="level2" data-number="10.6">
<h2 data-number="10.6" class="anchored" data-anchor-id="parentage-analysis"><span class="header-section-number">10.6</span> Parentage Analysis</h2>
<section id="mating-and-offspring" class="level3" data-number="10.6.1">
<h3 data-number="10.6.1" class="anchored" data-anchor-id="mating-and-offspring"><span class="header-section-number">10.6.1</span> Mating and Offspring</h3>
<p>The <code>mate()</code> function creates offspring from two parents:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>f <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">Locus =</span> <span class="fu">rep</span>(<span class="fu">paste0</span>(<span class="st">"Loc"</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>), <span class="at">each =</span> <span class="dv">3</span>),</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">Allele =</span> <span class="fu">rep</span>(LETTERS[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>], <span class="at">times =</span> <span class="dv">4</span>),</span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">Frequency =</span> <span class="fu">rep</span>(<span class="fu">c</span>(<span class="dv">1</span><span class="sc">/</span><span class="dv">3</span>, <span class="dv">1</span><span class="sc">/</span><span class="dv">3</span>, <span class="dv">1</span><span class="sc">/</span><span class="dv">3</span>), <span class="at">times =</span> <span class="dv">4</span>)</span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a>adults <span class="ot">&lt;-</span> <span class="fu">make_population</span>(f, <span class="at">N =</span> <span class="dv">20</span>)</span>
<span id="cb30-7"><a href="#cb30-7" aria-hidden="true" tabindex="-1"></a>offs <span class="ot">&lt;-</span> <span class="fu">mate</span>(adults[<span class="dv">1</span>, ], adults[<span class="dv">2</span>, ], <span class="at">N =</span> <span class="dv">5</span>)</span>
<span id="cb30-8"><a href="#cb30-8" aria-hidden="true" tabindex="-1"></a>offs</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>  ID Loc1 Loc2 Loc3 Loc4
1  1  A:C  C:C  B:C  B:C
2  1  A:A  A:C  B:C  B:C
3  1  B:C  C:C  B:C  B:C
4  1  A:C  A:C  B:B  B:C
5  1  A:B  A:C  B:C  B:C</code></pre>
</div>
</div>
</section>
<section id="maternal-subtraction" class="level3" data-number="10.6.2">
<h3 data-number="10.6.2" class="anchored" data-anchor-id="maternal-subtraction"><span class="header-section-number">10.6.2</span> Maternal Subtraction</h3>
<p>The <code>minus_mom()</code> function removes the maternal contribution from offspring genotypes. It takes a single data.frame where mothers have <code>OffID = 0</code> and offspring have <code>OffID != 0</code>:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Combine mother and offspring into one data.frame</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>mom_row <span class="ot">&lt;-</span> adults[<span class="dv">1</span>, ]</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>mom_row<span class="sc">$</span>OffID <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a>offs<span class="sc">$</span>OffID <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(offs)</span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a>family <span class="ot">&lt;-</span> <span class="fu">rbind</span>(mom_row, offs)</span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a>paternal <span class="ot">&lt;-</span> <span class="fu">minus_mom</span>(family)</span>
<span id="cb32-7"><a href="#cb32-7" aria-hidden="true" tabindex="-1"></a>paternal</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>  ID Loc1 Loc2 Loc3 Loc4 OffID
2  1  A:C    C  B:C    C     1
3  1    A  A:C  B:C    C     2
4  1    B    C  B:C    C     3
5  1  A:C  A:C    B    C     4
6  1    B  A:C  B:C    C     5</code></pre>
</div>
</div>
</section>
<section id="fractional-paternity" class="level3" data-number="10.6.3">
<h3 data-number="10.6.3" class="anchored" data-anchor-id="fractional-paternity"><span class="header-section-number">10.6.3</span> Fractional Paternity</h3>
<p>The <code>paternity()</code> function estimates the probability of each potential father siring each offspring:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>offs<span class="sc">$</span>OffID <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(offs)</span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>offs<span class="sc">$</span>MomID <span class="ot">&lt;-</span> adults<span class="sc">$</span>ID[<span class="dv">1</span>]</span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>pat <span class="ot">&lt;-</span> <span class="fu">paternity</span>(offs, adults[<span class="dv">1</span>, ], adults)</span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(pat)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>  MomID OffID DadID        Fij
1     1     1     2 0.29629630
2     1     1     7 0.14814815
3     1     1    12 0.14814815
4     1     1    15 0.14814815
5     1     1    18 0.14814815
6     1     1    16 0.07407407</code></pre>
</div>
</div>
<p>The <code>Fij</code> column gives the fractional paternity probability. Use <code>strict = TRUE</code> to retain only unambiguous assignments:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>pat_strict <span class="ot">&lt;-</span> <span class="fu">paternity</span>(offs, adults[<span class="dv">1</span>, ], adults, <span class="at">strict =</span> <span class="cn">TRUE</span>)</span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>pat_strict</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] MomID OffID DadID Fij  
&lt;0 rows&gt; (or 0-length row.names)</code></pre>
</div>
</div>
</section>
<section id="single-parent-exclusion" class="level3" data-number="10.6.4">
<h3 data-number="10.6.4" class="anchored" data-anchor-id="single-parent-exclusion"><span class="header-section-number">10.6.4</span> Single-Parent Exclusion</h3>
<p>The <code>parent_finder()</code> function identifies compatible parents for each offspring using exclusion. It requires a data.frame with <code>ID</code> and <code>OffID</code> columns, where adults have <code>OffID = 0</code>:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Prepare a combined data.frame</span></span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>locus_cols <span class="ot">&lt;-</span> <span class="fu">column_class</span>(adults, <span class="st">"locus"</span>)</span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>adults2 <span class="ot">&lt;-</span> adults[, <span class="fu">c</span>(<span class="st">"ID"</span>, locus_cols)]</span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a>adults2<span class="sc">$</span>OffID <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb38-5"><a href="#cb38-5" aria-hidden="true" tabindex="-1"></a>offs2 <span class="ot">&lt;-</span> offs[, <span class="fu">c</span>(<span class="st">"ID"</span>, locus_cols)]</span>
<span id="cb38-6"><a href="#cb38-6" aria-hidden="true" tabindex="-1"></a>offs2<span class="sc">$</span>OffID <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(offs2)</span>
<span id="cb38-7"><a href="#cb38-7" aria-hidden="true" tabindex="-1"></a>combined <span class="ot">&lt;-</span> <span class="fu">rbind</span>(adults2, offs2)</span>
<span id="cb38-8"><a href="#cb38-8" aria-hidden="true" tabindex="-1"></a>result <span class="ot">&lt;-</span> <span class="fu">parent_finder</span>(combined)</span>
<span id="cb38-9"><a href="#cb38-9" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(result)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>  ID OffID ParentID       T
1  1     1        1 0.06250
2  1     1        2 0.06250
4  1     1        7 0.06250
6  1     1       15 0.06250
3  1     1        5 0.03125
5  1     1       12 0.03125</code></pre>
</div>
</div>
</section>
<section id="identifying-bad-parents" class="level3" data-number="10.6.5">
<h3 data-number="10.6.5" class="anchored" data-anchor-id="identifying-bad-parents"><span class="header-section-number">10.6.5</span> Identifying Bad Parents</h3>
<p>The <code>bad_parents()</code> function flags parent-offspring pairs that are genetically incompatible:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="fu">bad_parents</span>(offs, adults)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
</section>
<section id="bundled-parentage-datasets" class="level2" data-number="10.7">
<h2 data-number="10.7" class="anchored" data-anchor-id="bundled-parentage-datasets"><span class="header-section-number">10.7</span> Bundled Parentage Datasets</h2>
<p>The <code>cornus</code> and <code>cornus_florida</code> datasets contain <em>Cornus florida</em> genotype data for parentage analysis:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(cornus)</span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(cornus)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Population SampleID X.Coordinate Y.Coordinate   Cf.G8  Cf.H18   Cf.N5  Cf.N10
1          2      203         3040         6146 157:165 117:119 170:170 189:191
2          2      204         3070         6114 159:163 101:103 166:170 193:201
3          2      205         3096         6148 151:157 113:119         193:195
4          2      206         3124         6136 153:161 117:119         197:201
5          2      207         3082         6126 159:163 105:119 158:170 193:193
6          2      208         3148         5180 181:181 119:119 170:170 199:201
    Cf.O5
1 198:198
2 178:182
3 178:182
4 178:198
5 178:196
6 176:182</code></pre>
</div>
</div>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(cornus_florida)</span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(cornus_florida)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>   ID OffID    X    Y      G8     H18      N5     N10      O5
1 226     0 1392 3534 162:180 114:114 124:126 192:192 185:195
2 232     0 1656 3414 158:180 112:112 124:126 184:192 185:185
3 234     0 1718 3330 158:180  112:96 128:128 184:192 185:185
4 300     0 1175 3114 180:188 112:116 126:126 198:200 191:195
5 305     0 1529 3237 154:170 122:124 124:126 188:192 181:195
6 432     0 1336 2748 164:180 114:116 124:126 198:202 185:193</code></pre>
</div>
</div>


<!-- -->

</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "î§‹";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
      const outerScaffold = trigger.parentElement.cloneNode(true);
      const codeEl = outerScaffold.querySelector('code');
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
    const viewSource = window.document.getElementById('quarto-view-source') ||
                       window.document.getElementById('quarto-code-tools-source');
    if (viewSource) {
      const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
      viewSource.addEventListener("click", function(e) {
        if (sourceUrl) {
          // rstudio viewer pane
          if (/\bcapabilities=\b/.test(window.location)) {
            window.open(sourceUrl);
          } else {
            window.location.href = sourceUrl;
          }
        } else {
          const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
          modal.show();
        }
        return false;
      });
    }
    function toggleCodeHandler(show) {
      return function(e) {
        const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
        for (let i=0; i<detailsSrc.length; i++) {
          const details = detailsSrc[i].parentElement;
          if (show) {
            details.open = true;
          } else {
            details.removeAttribute("open");
          }
        }
        const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
        const fromCls = show ? "hidden" : "unhidden";
        const toCls = show ? "unhidden" : "hidden";
        for (let i=0; i<cellCodeDivs.length; i++) {
          const codeDiv = cellCodeDivs[i];
          if (codeDiv.classList.contains(fromCls)) {
            codeDiv.classList.remove(fromCls);
            codeDiv.classList.add(toCls);
          } 
        }
        return false;
      }
    }
    const hideAllCode = window.document.getElementById("quarto-hide-all-code");
    if (hideAllCode) {
      hideAllCode.addEventListener("click", toggleCodeHandler(false));
    }
    const showAllCode = window.document.getElementById("quarto-show-all-code");
    if (showAllCode) {
      showAllCode.addEventListener("click", toggleCodeHandler(true));
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./visualization.html" class="pagination-link" aria-label="Visualization">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Visualization</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
  </div>
</nav><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb45" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="fu"># Simulation &amp; Parentage</span></span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a>This chapter covers the forward-time population simulation framework and parentage analysis tools in <span class="in">`gstudio`</span>.</span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-7"><a href="#cb45-7" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-8"><a href="#cb45-8" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(gstudio)</span>
<span id="cb45-9"><a href="#cb45-9" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr, <span class="at">warn.conflicts =</span> <span class="cn">FALSE</span>)</span>
<span id="cb45-10"><a href="#cb45-10" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-11"><a href="#cb45-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-12"><a href="#cb45-12" aria-hidden="true" tabindex="-1"></a><span class="fu">## Building Populations from Frequencies</span></span>
<span id="cb45-13"><a href="#cb45-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-14"><a href="#cb45-14" aria-hidden="true" tabindex="-1"></a><span class="fu">### Allele Frequency Specification</span></span>
<span id="cb45-15"><a href="#cb45-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-16"><a href="#cb45-16" aria-hidden="true" tabindex="-1"></a>The simulation tools start with specifying allele frequencies per population and locus. The <span class="in">`make_populations()`</span> function requires a data.frame with <span class="in">`Population`</span>, <span class="in">`Locus`</span>, <span class="in">`Allele`</span>, and <span class="in">`Frequency`</span> columns:</span>
<span id="cb45-17"><a href="#cb45-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-20"><a href="#cb45-20" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-21"><a href="#cb45-21" aria-hidden="true" tabindex="-1"></a>f1 <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb45-22"><a href="#cb45-22" aria-hidden="true" tabindex="-1"></a>  <span class="at">Population =</span> <span class="st">"A"</span>,</span>
<span id="cb45-23"><a href="#cb45-23" aria-hidden="true" tabindex="-1"></a>  <span class="at">Locus =</span> <span class="fu">rep</span>(<span class="fu">c</span>(<span class="st">"Loc1"</span>, <span class="st">"Loc2"</span>, <span class="st">"Loc3"</span>), <span class="at">each =</span> <span class="dv">2</span>),</span>
<span id="cb45-24"><a href="#cb45-24" aria-hidden="true" tabindex="-1"></a>  <span class="at">Allele =</span> <span class="fu">rep</span>(<span class="fu">c</span>(<span class="st">"01"</span>, <span class="st">"02"</span>), <span class="at">times =</span> <span class="dv">3</span>),</span>
<span id="cb45-25"><a href="#cb45-25" aria-hidden="true" tabindex="-1"></a>  <span class="at">Frequency =</span> <span class="fu">c</span>(<span class="fl">0.2</span>, <span class="fl">0.8</span>, <span class="fl">0.3</span>, <span class="fl">0.7</span>, <span class="fl">0.4</span>, <span class="fl">0.6</span>)</span>
<span id="cb45-26"><a href="#cb45-26" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb45-27"><a href="#cb45-27" aria-hidden="true" tabindex="-1"></a>f2 <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb45-28"><a href="#cb45-28" aria-hidden="true" tabindex="-1"></a>  <span class="at">Population =</span> <span class="st">"B"</span>,</span>
<span id="cb45-29"><a href="#cb45-29" aria-hidden="true" tabindex="-1"></a>  <span class="at">Locus =</span> <span class="fu">rep</span>(<span class="fu">c</span>(<span class="st">"Loc1"</span>, <span class="st">"Loc2"</span>, <span class="st">"Loc3"</span>), <span class="at">each =</span> <span class="dv">2</span>),</span>
<span id="cb45-30"><a href="#cb45-30" aria-hidden="true" tabindex="-1"></a>  <span class="at">Allele =</span> <span class="fu">rep</span>(<span class="fu">c</span>(<span class="st">"01"</span>, <span class="st">"02"</span>), <span class="at">times =</span> <span class="dv">3</span>),</span>
<span id="cb45-31"><a href="#cb45-31" aria-hidden="true" tabindex="-1"></a>  <span class="at">Frequency =</span> <span class="fu">c</span>(<span class="fl">0.7</span>, <span class="fl">0.3</span>, <span class="fl">0.5</span>, <span class="fl">0.5</span>, <span class="fl">0.6</span>, <span class="fl">0.4</span>)</span>
<span id="cb45-32"><a href="#cb45-32" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb45-33"><a href="#cb45-33" aria-hidden="true" tabindex="-1"></a>f3 <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb45-34"><a href="#cb45-34" aria-hidden="true" tabindex="-1"></a>  <span class="at">Population =</span> <span class="st">"C"</span>,</span>
<span id="cb45-35"><a href="#cb45-35" aria-hidden="true" tabindex="-1"></a>  <span class="at">Locus =</span> <span class="fu">rep</span>(<span class="fu">c</span>(<span class="st">"Loc1"</span>, <span class="st">"Loc2"</span>, <span class="st">"Loc3"</span>), <span class="at">each =</span> <span class="dv">2</span>),</span>
<span id="cb45-36"><a href="#cb45-36" aria-hidden="true" tabindex="-1"></a>  <span class="at">Allele =</span> <span class="fu">rep</span>(<span class="fu">c</span>(<span class="st">"01"</span>, <span class="st">"02"</span>), <span class="at">times =</span> <span class="dv">3</span>),</span>
<span id="cb45-37"><a href="#cb45-37" aria-hidden="true" tabindex="-1"></a>  <span class="at">Frequency =</span> <span class="fu">c</span>(<span class="fl">0.4</span>, <span class="fl">0.6</span>, <span class="fl">0.3</span>, <span class="fl">0.7</span>, <span class="fl">0.2</span>, <span class="fl">0.8</span>)</span>
<span id="cb45-38"><a href="#cb45-38" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb45-39"><a href="#cb45-39" aria-hidden="true" tabindex="-1"></a>freqs <span class="ot">&lt;-</span> <span class="fu">rbind</span>(f1, f2, f3)</span>
<span id="cb45-40"><a href="#cb45-40" aria-hidden="true" tabindex="-1"></a>freqs<span class="sc">$</span>Population <span class="ot">&lt;-</span> <span class="fu">factor</span>(freqs<span class="sc">$</span>Population, <span class="at">ordered =</span> <span class="cn">TRUE</span>)</span>
<span id="cb45-41"><a href="#cb45-41" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-42"><a href="#cb45-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-43"><a href="#cb45-43" aria-hidden="true" tabindex="-1"></a><span class="fu">### Creating Populations</span></span>
<span id="cb45-44"><a href="#cb45-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-47"><a href="#cb45-47" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-48"><a href="#cb45-48" aria-hidden="true" tabindex="-1"></a>pop <span class="ot">&lt;-</span> <span class="fu">make_populations</span>(freqs, <span class="at">N =</span> <span class="dv">30</span>)</span>
<span id="cb45-49"><a href="#cb45-49" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(pop)</span>
<span id="cb45-50"><a href="#cb45-50" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(pop<span class="sc">$</span>Population)</span>
<span id="cb45-51"><a href="#cb45-51" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-52"><a href="#cb45-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-53"><a href="#cb45-53" aria-hidden="true" tabindex="-1"></a>You can specify an inbreeding level with <span class="in">`F`</span>:</span>
<span id="cb45-54"><a href="#cb45-54" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-57"><a href="#cb45-57" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-58"><a href="#cb45-58" aria-hidden="true" tabindex="-1"></a>pop_inbred <span class="ot">&lt;-</span> <span class="fu">make_populations</span>(freqs, <span class="at">N =</span> <span class="dv">30</span>, <span class="at">F =</span> <span class="fl">0.3</span>)</span>
<span id="cb45-59"><a href="#cb45-59" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-60"><a href="#cb45-60" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-61"><a href="#cb45-61" aria-hidden="true" tabindex="-1"></a><span class="fu">## Migration Models</span></span>
<span id="cb45-62"><a href="#cb45-62" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-63"><a href="#cb45-63" aria-hidden="true" tabindex="-1"></a>The <span class="in">`migration_matrix()`</span> function creates a K x K migration matrix where rows sum to 1. Element <span class="in">`[i,j]`</span> is the proportion of population *i* that originates from population *j*.</span>
<span id="cb45-64"><a href="#cb45-64" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-65"><a href="#cb45-65" aria-hidden="true" tabindex="-1"></a><span class="fu">### Island Model</span></span>
<span id="cb45-66"><a href="#cb45-66" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-67"><a href="#cb45-67" aria-hidden="true" tabindex="-1"></a>Equal migration among all populations:</span>
<span id="cb45-68"><a href="#cb45-68" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-71"><a href="#cb45-71" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-72"><a href="#cb45-72" aria-hidden="true" tabindex="-1"></a><span class="fu">migration_matrix</span>(<span class="dv">3</span>, <span class="at">model =</span> <span class="st">"island"</span>, <span class="at">m =</span> <span class="fl">0.05</span>)</span>
<span id="cb45-73"><a href="#cb45-73" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-74"><a href="#cb45-74" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-75"><a href="#cb45-75" aria-hidden="true" tabindex="-1"></a><span class="fu">### 1D Stepping Stone</span></span>
<span id="cb45-76"><a href="#cb45-76" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-77"><a href="#cb45-77" aria-hidden="true" tabindex="-1"></a>Migration only between adjacent populations:</span>
<span id="cb45-78"><a href="#cb45-78" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-81"><a href="#cb45-81" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-82"><a href="#cb45-82" aria-hidden="true" tabindex="-1"></a><span class="fu">migration_matrix</span>(<span class="fu">c</span>(<span class="st">"A"</span>, <span class="st">"B"</span>, <span class="st">"C"</span>, <span class="st">"D"</span>),</span>
<span id="cb45-83"><a href="#cb45-83" aria-hidden="true" tabindex="-1"></a>                 <span class="at">model =</span> <span class="st">"stepping_stone_1d"</span>, <span class="at">m =</span> <span class="fl">0.1</span>)</span>
<span id="cb45-84"><a href="#cb45-84" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-85"><a href="#cb45-85" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-86"><a href="#cb45-86" aria-hidden="true" tabindex="-1"></a><span class="fu">### 2D Stepping Stone</span></span>
<span id="cb45-87"><a href="#cb45-87" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-88"><a href="#cb45-88" aria-hidden="true" tabindex="-1"></a>Grid-based adjacency:</span>
<span id="cb45-89"><a href="#cb45-89" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-92"><a href="#cb45-92" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-93"><a href="#cb45-93" aria-hidden="true" tabindex="-1"></a><span class="fu">migration_matrix</span>(<span class="dv">4</span>, <span class="at">model =</span> <span class="st">"stepping_stone_2d"</span>,</span>
<span id="cb45-94"><a href="#cb45-94" aria-hidden="true" tabindex="-1"></a>                 <span class="at">m =</span> <span class="fl">0.1</span>, <span class="at">nr =</span> <span class="dv">2</span>, <span class="at">nc =</span> <span class="dv">2</span>)</span>
<span id="cb45-95"><a href="#cb45-95" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-96"><a href="#cb45-96" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-97"><a href="#cb45-97" aria-hidden="true" tabindex="-1"></a><span class="fu">### Distance-Based</span></span>
<span id="cb45-98"><a href="#cb45-98" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-99"><a href="#cb45-99" aria-hidden="true" tabindex="-1"></a>Migration rate decays with distance:</span>
<span id="cb45-100"><a href="#cb45-100" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-103"><a href="#cb45-103" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-104"><a href="#cb45-104" aria-hidden="true" tabindex="-1"></a>coords <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">1</span>, <span class="dv">0</span>, <span class="fl">0.5</span>, <span class="dv">1</span>), <span class="at">ncol =</span> <span class="dv">2</span>, <span class="at">byrow =</span> <span class="cn">TRUE</span>)</span>
<span id="cb45-105"><a href="#cb45-105" aria-hidden="true" tabindex="-1"></a><span class="fu">migration_matrix</span>(<span class="dv">3</span>, <span class="at">model =</span> <span class="st">"distance"</span>, <span class="at">m =</span> <span class="fl">0.1</span>, <span class="at">coords =</span> coords)</span>
<span id="cb45-106"><a href="#cb45-106" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-107"><a href="#cb45-107" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-108"><a href="#cb45-108" aria-hidden="true" tabindex="-1"></a><span class="fu">## Migration Events</span></span>
<span id="cb45-109"><a href="#cb45-109" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-110"><a href="#cb45-110" aria-hidden="true" tabindex="-1"></a>A <span class="in">`migration_event`</span> pairs a migration matrix with a generation interval, allowing temporal regime changes:</span>
<span id="cb45-111"><a href="#cb45-111" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-114"><a href="#cb45-114" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-115"><a href="#cb45-115" aria-hidden="true" tabindex="-1"></a>m1 <span class="ot">&lt;-</span> <span class="fu">migration_matrix</span>(<span class="dv">3</span>, <span class="at">model =</span> <span class="st">"island"</span>, <span class="at">m =</span> <span class="fl">0.05</span>)</span>
<span id="cb45-116"><a href="#cb45-116" aria-hidden="true" tabindex="-1"></a>m2 <span class="ot">&lt;-</span> <span class="fu">migration_matrix</span>(<span class="dv">3</span>, <span class="at">model =</span> <span class="st">"island"</span>, <span class="at">m =</span> <span class="fl">0.20</span>)</span>
<span id="cb45-117"><a href="#cb45-117" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-118"><a href="#cb45-118" aria-hidden="true" tabindex="-1"></a><span class="co"># Low migration for generations 1-50</span></span>
<span id="cb45-119"><a href="#cb45-119" aria-hidden="true" tabindex="-1"></a>evt1 <span class="ot">&lt;-</span> <span class="fu">migration_event</span>(m1, <span class="at">start =</span> <span class="dv">1</span>, <span class="at">end =</span> <span class="dv">50</span>)</span>
<span id="cb45-120"><a href="#cb45-120" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-121"><a href="#cb45-121" aria-hidden="true" tabindex="-1"></a><span class="co"># High migration from generation 51 onward</span></span>
<span id="cb45-122"><a href="#cb45-122" aria-hidden="true" tabindex="-1"></a>evt2 <span class="ot">&lt;-</span> <span class="fu">migration_event</span>(m2, <span class="at">start =</span> <span class="dv">51</span>)</span>
<span id="cb45-123"><a href="#cb45-123" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-124"><a href="#cb45-124" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-125"><a href="#cb45-125" aria-hidden="true" tabindex="-1"></a><span class="fu">## Mutation Models</span></span>
<span id="cb45-126"><a href="#cb45-126" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-127"><a href="#cb45-127" aria-hidden="true" tabindex="-1"></a>The <span class="in">`mutation_model()`</span> function defines how alleles mutate during simulation:</span>
<span id="cb45-128"><a href="#cb45-128" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-131"><a href="#cb45-131" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-132"><a href="#cb45-132" aria-hidden="true" tabindex="-1"></a><span class="co"># Infinite Alleles Model</span></span>
<span id="cb45-133"><a href="#cb45-133" aria-hidden="true" tabindex="-1"></a>mm_iam <span class="ot">&lt;-</span> <span class="fu">mutation_model</span>(<span class="at">rate =</span> <span class="fl">0.001</span>, <span class="at">model =</span> <span class="st">"iam"</span>)</span>
<span id="cb45-134"><a href="#cb45-134" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-135"><a href="#cb45-135" aria-hidden="true" tabindex="-1"></a><span class="co"># K-Allele Model (10 possible states)</span></span>
<span id="cb45-136"><a href="#cb45-136" aria-hidden="true" tabindex="-1"></a>mm_kam <span class="ot">&lt;-</span> <span class="fu">mutation_model</span>(<span class="at">rate =</span> <span class="fl">0.001</span>, <span class="at">model =</span> <span class="st">"kam"</span>, <span class="at">k =</span> <span class="dv">10</span>)</span>
<span id="cb45-137"><a href="#cb45-137" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-138"><a href="#cb45-138" aria-hidden="true" tabindex="-1"></a><span class="co"># Stepwise Mutation Model</span></span>
<span id="cb45-139"><a href="#cb45-139" aria-hidden="true" tabindex="-1"></a>mm_smm <span class="ot">&lt;-</span> <span class="fu">mutation_model</span>(<span class="at">rate =</span> <span class="fl">0.001</span>, <span class="at">model =</span> <span class="st">"smm"</span>)</span>
<span id="cb45-140"><a href="#cb45-140" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-141"><a href="#cb45-141" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-142"><a href="#cb45-142" aria-hidden="true" tabindex="-1"></a><span class="fu">## Forward-Time Simulation</span></span>
<span id="cb45-143"><a href="#cb45-143" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-144"><a href="#cb45-144" aria-hidden="true" tabindex="-1"></a>The <span class="in">`simulate_pop()`</span> function runs a multi-generation forward-time simulation. Each generation cycle: census, migrate, mate, mutate.</span>
<span id="cb45-145"><a href="#cb45-145" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-146"><a href="#cb45-146" aria-hidden="true" tabindex="-1"></a><span class="fu">### Basic Simulation</span></span>
<span id="cb45-147"><a href="#cb45-147" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-150"><a href="#cb45-150" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-151"><a href="#cb45-151" aria-hidden="true" tabindex="-1"></a>pop <span class="ot">&lt;-</span> <span class="fu">make_populations</span>(freqs, <span class="at">N =</span> <span class="dv">30</span>)</span>
<span id="cb45-152"><a href="#cb45-152" aria-hidden="true" tabindex="-1"></a>result <span class="ot">&lt;-</span> <span class="fu">simulate_pop</span>(pop, <span class="at">ngen =</span> <span class="dv">20</span>, <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb45-153"><a href="#cb45-153" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(result)</span>
<span id="cb45-154"><a href="#cb45-154" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-155"><a href="#cb45-155" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-156"><a href="#cb45-156" aria-hidden="true" tabindex="-1"></a><span class="fu">### With Migration</span></span>
<span id="cb45-157"><a href="#cb45-157" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-160"><a href="#cb45-160" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-161"><a href="#cb45-161" aria-hidden="true" tabindex="-1"></a>m <span class="ot">&lt;-</span> <span class="fu">migration_matrix</span>(<span class="dv">3</span>, <span class="at">model =</span> <span class="st">"island"</span>, <span class="at">m =</span> <span class="fl">0.05</span>)</span>
<span id="cb45-162"><a href="#cb45-162" aria-hidden="true" tabindex="-1"></a>evt <span class="ot">&lt;-</span> <span class="fu">migration_event</span>(m, <span class="at">start =</span> <span class="dv">1</span>)</span>
<span id="cb45-163"><a href="#cb45-163" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-164"><a href="#cb45-164" aria-hidden="true" tabindex="-1"></a>result <span class="ot">&lt;-</span> <span class="fu">simulate_pop</span>(pop, <span class="at">ngen =</span> <span class="dv">20</span>,</span>
<span id="cb45-165"><a href="#cb45-165" aria-hidden="true" tabindex="-1"></a>                       <span class="at">migration =</span> evt,</span>
<span id="cb45-166"><a href="#cb45-166" aria-hidden="true" tabindex="-1"></a>                       <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb45-167"><a href="#cb45-167" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-168"><a href="#cb45-168" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-169"><a href="#cb45-169" aria-hidden="true" tabindex="-1"></a><span class="fu">### With Mutation</span></span>
<span id="cb45-170"><a href="#cb45-170" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-173"><a href="#cb45-173" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-174"><a href="#cb45-174" aria-hidden="true" tabindex="-1"></a>mm <span class="ot">&lt;-</span> <span class="fu">mutation_model</span>(<span class="at">rate =</span> <span class="fl">0.01</span>, <span class="at">model =</span> <span class="st">"iam"</span>)</span>
<span id="cb45-175"><a href="#cb45-175" aria-hidden="true" tabindex="-1"></a>result <span class="ot">&lt;-</span> <span class="fu">simulate_pop</span>(pop, <span class="at">ngen =</span> <span class="dv">20</span>,</span>
<span id="cb45-176"><a href="#cb45-176" aria-hidden="true" tabindex="-1"></a>                       <span class="at">mutation =</span> mm,</span>
<span id="cb45-177"><a href="#cb45-177" aria-hidden="true" tabindex="-1"></a>                       <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb45-178"><a href="#cb45-178" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-179"><a href="#cb45-179" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-180"><a href="#cb45-180" aria-hidden="true" tabindex="-1"></a><span class="fu">### With Mixed Mating (Selfing)</span></span>
<span id="cb45-181"><a href="#cb45-181" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-184"><a href="#cb45-184" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-185"><a href="#cb45-185" aria-hidden="true" tabindex="-1"></a>result <span class="ot">&lt;-</span> <span class="fu">simulate_pop</span>(pop, <span class="at">ngen =</span> <span class="dv">20</span>,</span>
<span id="cb45-186"><a href="#cb45-186" aria-hidden="true" tabindex="-1"></a>                       <span class="at">selfing_rate =</span> <span class="fl">0.3</span>,</span>
<span id="cb45-187"><a href="#cb45-187" aria-hidden="true" tabindex="-1"></a>                       <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb45-188"><a href="#cb45-188" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-189"><a href="#cb45-189" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-190"><a href="#cb45-190" aria-hidden="true" tabindex="-1"></a><span class="fu">### Census Snapshots</span></span>
<span id="cb45-191"><a href="#cb45-191" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-192"><a href="#cb45-192" aria-hidden="true" tabindex="-1"></a>Save population snapshots at regular intervals:</span>
<span id="cb45-193"><a href="#cb45-193" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-196"><a href="#cb45-196" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-197"><a href="#cb45-197" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb45-198"><a href="#cb45-198" aria-hidden="true" tabindex="-1"></a>result <span class="ot">&lt;-</span> <span class="fu">simulate_pop</span>(pop, <span class="at">ngen =</span> <span class="dv">100</span>,</span>
<span id="cb45-199"><a href="#cb45-199" aria-hidden="true" tabindex="-1"></a>                       <span class="at">census_interval =</span> <span class="dv">25</span>,</span>
<span id="cb45-200"><a href="#cb45-200" aria-hidden="true" tabindex="-1"></a>                       <span class="at">census_dir =</span> <span class="fu">tempdir</span>(),</span>
<span id="cb45-201"><a href="#cb45-201" aria-hidden="true" tabindex="-1"></a>                       <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb45-202"><a href="#cb45-202" aria-hidden="true" tabindex="-1"></a><span class="co"># Snapshots saved as gen_0025.rds, gen_0050.rds, etc.</span></span>
<span id="cb45-203"><a href="#cb45-203" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-204"><a href="#cb45-204" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-205"><a href="#cb45-205" aria-hidden="true" tabindex="-1"></a><span class="fu">### Complete Worked Example</span></span>
<span id="cb45-206"><a href="#cb45-206" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-207"><a href="#cb45-207" aria-hidden="true" tabindex="-1"></a>Simulate drift and migration in a 3-population system, then analyze the outcome:</span>
<span id="cb45-208"><a href="#cb45-208" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-211"><a href="#cb45-211" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-212"><a href="#cb45-212" aria-hidden="true" tabindex="-1"></a><span class="co"># Create starting populations</span></span>
<span id="cb45-213"><a href="#cb45-213" aria-hidden="true" tabindex="-1"></a>pop <span class="ot">&lt;-</span> <span class="fu">make_populations</span>(freqs, <span class="at">N =</span> <span class="dv">50</span>)</span>
<span id="cb45-214"><a href="#cb45-214" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-215"><a href="#cb45-215" aria-hidden="true" tabindex="-1"></a><span class="co"># Set up island model migration</span></span>
<span id="cb45-216"><a href="#cb45-216" aria-hidden="true" tabindex="-1"></a>m <span class="ot">&lt;-</span> <span class="fu">migration_matrix</span>(<span class="dv">3</span>, <span class="at">model =</span> <span class="st">"island"</span>, <span class="at">m =</span> <span class="fl">0.02</span>)</span>
<span id="cb45-217"><a href="#cb45-217" aria-hidden="true" tabindex="-1"></a>evt <span class="ot">&lt;-</span> <span class="fu">migration_event</span>(m, <span class="at">start =</span> <span class="dv">1</span>)</span>
<span id="cb45-218"><a href="#cb45-218" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-219"><a href="#cb45-219" aria-hidden="true" tabindex="-1"></a><span class="co"># Run simulation</span></span>
<span id="cb45-220"><a href="#cb45-220" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">42</span>)</span>
<span id="cb45-221"><a href="#cb45-221" aria-hidden="true" tabindex="-1"></a>final <span class="ot">&lt;-</span> <span class="fu">simulate_pop</span>(pop, <span class="at">ngen =</span> <span class="dv">50</span>,</span>
<span id="cb45-222"><a href="#cb45-222" aria-hidden="true" tabindex="-1"></a>                      <span class="at">migration =</span> evt,</span>
<span id="cb45-223"><a href="#cb45-223" aria-hidden="true" tabindex="-1"></a>                      <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb45-224"><a href="#cb45-224" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-225"><a href="#cb45-225" aria-hidden="true" tabindex="-1"></a><span class="co"># Analyze structure in the final generation</span></span>
<span id="cb45-226"><a href="#cb45-226" aria-hidden="true" tabindex="-1"></a><span class="fu">genetic_structure</span>(final, <span class="at">stratum =</span> <span class="st">"Population"</span>, <span class="at">mode =</span> <span class="st">"Gst"</span>)</span>
<span id="cb45-227"><a href="#cb45-227" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-228"><a href="#cb45-228" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-231"><a href="#cb45-231" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-232"><a href="#cb45-232" aria-hidden="true" tabindex="-1"></a><span class="co"># Compare diversity before and after</span></span>
<span id="cb45-233"><a href="#cb45-233" aria-hidden="true" tabindex="-1"></a><span class="fu">genetic_diversity</span>(pop, <span class="at">stratum =</span> <span class="st">"Population"</span>, <span class="at">mode =</span> <span class="st">"He"</span>)</span>
<span id="cb45-234"><a href="#cb45-234" aria-hidden="true" tabindex="-1"></a><span class="fu">genetic_diversity</span>(final, <span class="at">stratum =</span> <span class="st">"Population"</span>, <span class="at">mode =</span> <span class="st">"He"</span>)</span>
<span id="cb45-235"><a href="#cb45-235" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-236"><a href="#cb45-236" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-237"><a href="#cb45-237" aria-hidden="true" tabindex="-1"></a><span class="fu">## Parentage Analysis</span></span>
<span id="cb45-238"><a href="#cb45-238" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-239"><a href="#cb45-239" aria-hidden="true" tabindex="-1"></a><span class="fu">### Mating and Offspring</span></span>
<span id="cb45-240"><a href="#cb45-240" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-241"><a href="#cb45-241" aria-hidden="true" tabindex="-1"></a>The <span class="in">`mate()`</span> function creates offspring from two parents:</span>
<span id="cb45-242"><a href="#cb45-242" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-245"><a href="#cb45-245" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-246"><a href="#cb45-246" aria-hidden="true" tabindex="-1"></a>f <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb45-247"><a href="#cb45-247" aria-hidden="true" tabindex="-1"></a>  <span class="at">Locus =</span> <span class="fu">rep</span>(<span class="fu">paste0</span>(<span class="st">"Loc"</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>), <span class="at">each =</span> <span class="dv">3</span>),</span>
<span id="cb45-248"><a href="#cb45-248" aria-hidden="true" tabindex="-1"></a>  <span class="at">Allele =</span> <span class="fu">rep</span>(LETTERS[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>], <span class="at">times =</span> <span class="dv">4</span>),</span>
<span id="cb45-249"><a href="#cb45-249" aria-hidden="true" tabindex="-1"></a>  <span class="at">Frequency =</span> <span class="fu">rep</span>(<span class="fu">c</span>(<span class="dv">1</span><span class="sc">/</span><span class="dv">3</span>, <span class="dv">1</span><span class="sc">/</span><span class="dv">3</span>, <span class="dv">1</span><span class="sc">/</span><span class="dv">3</span>), <span class="at">times =</span> <span class="dv">4</span>)</span>
<span id="cb45-250"><a href="#cb45-250" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb45-251"><a href="#cb45-251" aria-hidden="true" tabindex="-1"></a>adults <span class="ot">&lt;-</span> <span class="fu">make_population</span>(f, <span class="at">N =</span> <span class="dv">20</span>)</span>
<span id="cb45-252"><a href="#cb45-252" aria-hidden="true" tabindex="-1"></a>offs <span class="ot">&lt;-</span> <span class="fu">mate</span>(adults[<span class="dv">1</span>, ], adults[<span class="dv">2</span>, ], <span class="at">N =</span> <span class="dv">5</span>)</span>
<span id="cb45-253"><a href="#cb45-253" aria-hidden="true" tabindex="-1"></a>offs</span>
<span id="cb45-254"><a href="#cb45-254" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-255"><a href="#cb45-255" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-256"><a href="#cb45-256" aria-hidden="true" tabindex="-1"></a><span class="fu">### Maternal Subtraction</span></span>
<span id="cb45-257"><a href="#cb45-257" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-258"><a href="#cb45-258" aria-hidden="true" tabindex="-1"></a>The <span class="in">`minus_mom()`</span> function removes the maternal contribution from offspring genotypes. It takes a single data.frame where mothers have <span class="in">`OffID = 0`</span> and offspring have <span class="in">`OffID != 0`</span>:</span>
<span id="cb45-259"><a href="#cb45-259" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-262"><a href="#cb45-262" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-263"><a href="#cb45-263" aria-hidden="true" tabindex="-1"></a><span class="co"># Combine mother and offspring into one data.frame</span></span>
<span id="cb45-264"><a href="#cb45-264" aria-hidden="true" tabindex="-1"></a>mom_row <span class="ot">&lt;-</span> adults[<span class="dv">1</span>, ]</span>
<span id="cb45-265"><a href="#cb45-265" aria-hidden="true" tabindex="-1"></a>mom_row<span class="sc">$</span>OffID <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb45-266"><a href="#cb45-266" aria-hidden="true" tabindex="-1"></a>offs<span class="sc">$</span>OffID <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(offs)</span>
<span id="cb45-267"><a href="#cb45-267" aria-hidden="true" tabindex="-1"></a>family <span class="ot">&lt;-</span> <span class="fu">rbind</span>(mom_row, offs)</span>
<span id="cb45-268"><a href="#cb45-268" aria-hidden="true" tabindex="-1"></a>paternal <span class="ot">&lt;-</span> <span class="fu">minus_mom</span>(family)</span>
<span id="cb45-269"><a href="#cb45-269" aria-hidden="true" tabindex="-1"></a>paternal</span>
<span id="cb45-270"><a href="#cb45-270" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-271"><a href="#cb45-271" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-272"><a href="#cb45-272" aria-hidden="true" tabindex="-1"></a><span class="fu">### Fractional Paternity</span></span>
<span id="cb45-273"><a href="#cb45-273" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-274"><a href="#cb45-274" aria-hidden="true" tabindex="-1"></a>The <span class="in">`paternity()`</span> function estimates the probability of each potential father siring each offspring:</span>
<span id="cb45-275"><a href="#cb45-275" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-278"><a href="#cb45-278" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-279"><a href="#cb45-279" aria-hidden="true" tabindex="-1"></a>offs<span class="sc">$</span>OffID <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(offs)</span>
<span id="cb45-280"><a href="#cb45-280" aria-hidden="true" tabindex="-1"></a>offs<span class="sc">$</span>MomID <span class="ot">&lt;-</span> adults<span class="sc">$</span>ID[<span class="dv">1</span>]</span>
<span id="cb45-281"><a href="#cb45-281" aria-hidden="true" tabindex="-1"></a>pat <span class="ot">&lt;-</span> <span class="fu">paternity</span>(offs, adults[<span class="dv">1</span>, ], adults)</span>
<span id="cb45-282"><a href="#cb45-282" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(pat)</span>
<span id="cb45-283"><a href="#cb45-283" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-284"><a href="#cb45-284" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-285"><a href="#cb45-285" aria-hidden="true" tabindex="-1"></a>The <span class="in">`Fij`</span> column gives the fractional paternity probability. Use <span class="in">`strict = TRUE`</span> to retain only unambiguous assignments:</span>
<span id="cb45-286"><a href="#cb45-286" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-289"><a href="#cb45-289" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-290"><a href="#cb45-290" aria-hidden="true" tabindex="-1"></a>pat_strict <span class="ot">&lt;-</span> <span class="fu">paternity</span>(offs, adults[<span class="dv">1</span>, ], adults, <span class="at">strict =</span> <span class="cn">TRUE</span>)</span>
<span id="cb45-291"><a href="#cb45-291" aria-hidden="true" tabindex="-1"></a>pat_strict</span>
<span id="cb45-292"><a href="#cb45-292" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-293"><a href="#cb45-293" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-294"><a href="#cb45-294" aria-hidden="true" tabindex="-1"></a><span class="fu">### Single-Parent Exclusion</span></span>
<span id="cb45-295"><a href="#cb45-295" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-296"><a href="#cb45-296" aria-hidden="true" tabindex="-1"></a>The <span class="in">`parent_finder()`</span> function identifies compatible parents for each offspring using exclusion. It requires a data.frame with <span class="in">`ID`</span> and <span class="in">`OffID`</span> columns, where adults have <span class="in">`OffID = 0`</span>:</span>
<span id="cb45-297"><a href="#cb45-297" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-300"><a href="#cb45-300" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-301"><a href="#cb45-301" aria-hidden="true" tabindex="-1"></a><span class="co"># Prepare a combined data.frame</span></span>
<span id="cb45-302"><a href="#cb45-302" aria-hidden="true" tabindex="-1"></a>locus_cols <span class="ot">&lt;-</span> <span class="fu">column_class</span>(adults, <span class="st">"locus"</span>)</span>
<span id="cb45-303"><a href="#cb45-303" aria-hidden="true" tabindex="-1"></a>adults2 <span class="ot">&lt;-</span> adults[, <span class="fu">c</span>(<span class="st">"ID"</span>, locus_cols)]</span>
<span id="cb45-304"><a href="#cb45-304" aria-hidden="true" tabindex="-1"></a>adults2<span class="sc">$</span>OffID <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb45-305"><a href="#cb45-305" aria-hidden="true" tabindex="-1"></a>offs2 <span class="ot">&lt;-</span> offs[, <span class="fu">c</span>(<span class="st">"ID"</span>, locus_cols)]</span>
<span id="cb45-306"><a href="#cb45-306" aria-hidden="true" tabindex="-1"></a>offs2<span class="sc">$</span>OffID <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(offs2)</span>
<span id="cb45-307"><a href="#cb45-307" aria-hidden="true" tabindex="-1"></a>combined <span class="ot">&lt;-</span> <span class="fu">rbind</span>(adults2, offs2)</span>
<span id="cb45-308"><a href="#cb45-308" aria-hidden="true" tabindex="-1"></a>result <span class="ot">&lt;-</span> <span class="fu">parent_finder</span>(combined)</span>
<span id="cb45-309"><a href="#cb45-309" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(result)</span>
<span id="cb45-310"><a href="#cb45-310" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-311"><a href="#cb45-311" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-312"><a href="#cb45-312" aria-hidden="true" tabindex="-1"></a><span class="fu">### Identifying Bad Parents</span></span>
<span id="cb45-313"><a href="#cb45-313" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-314"><a href="#cb45-314" aria-hidden="true" tabindex="-1"></a>The <span class="in">`bad_parents()`</span> function flags parent-offspring pairs that are genetically incompatible:</span>
<span id="cb45-315"><a href="#cb45-315" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-318"><a href="#cb45-318" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-319"><a href="#cb45-319" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb45-320"><a href="#cb45-320" aria-hidden="true" tabindex="-1"></a><span class="fu">bad_parents</span>(offs, adults)</span>
<span id="cb45-321"><a href="#cb45-321" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-322"><a href="#cb45-322" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-323"><a href="#cb45-323" aria-hidden="true" tabindex="-1"></a><span class="fu">## Bundled Parentage Datasets</span></span>
<span id="cb45-324"><a href="#cb45-324" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-325"><a href="#cb45-325" aria-hidden="true" tabindex="-1"></a>The <span class="in">`cornus`</span> and <span class="in">`cornus_florida`</span> datasets contain *Cornus florida* genotype data for parentage analysis:</span>
<span id="cb45-326"><a href="#cb45-326" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-329"><a href="#cb45-329" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-330"><a href="#cb45-330" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(cornus)</span>
<span id="cb45-331"><a href="#cb45-331" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(cornus)</span>
<span id="cb45-332"><a href="#cb45-332" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-333"><a href="#cb45-333" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-336"><a href="#cb45-336" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-337"><a href="#cb45-337" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(cornus_florida)</span>
<span id="cb45-338"><a href="#cb45-338" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(cornus_florida)</span>
<span id="cb45-339"><a href="#cb45-339" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
</code></pre></div><button title="Copy to Clipboard" class="code-copy-button" data-in-quarto-modal=""><i class="bi"></i></button></div>
</div></div></div></div></div>
</div> <!-- /content -->




<footer class="footer"><div class="nav-footer"><div class="nav-footer-center"><div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/dyerlab/gstudio/issues/new" class="toc-action"><i class="bi bi-github"></i>Report an issue</a></li></ul></div></div></div></footer></body></html>